mkdocs/nav.py

```diff
@@ -14,18 +14,33 @@
 log = logging.getLogger(__name__)
 
 
-def filename_to_title(filename):
+def file_to_title(filename):
     """
     Automatically generate a default title, given a filename.
     """
     if utils.is_homepage(filename):
         return 'Home'
 
+    table_of_contents = None
+    try:
+        with open(filename, 'r') as f:
+            lines = f.read()
+            html_content, table_of_contents, meta = utils.convert_markdown(lines, ['toc'])
+            if len(table_of_contents.items) > 0:
```

Comment:
Shouldn't the meta-data be checked first? Perhaps the document author defined the title there:

    title: Some Document Title

    # A header that is not the document title.

Then, if the title is not defined in the meta-data, fall back to the first header in the document.